tests/source_encapsulation/build.bp

```diff
@@ -0,0 +1,134 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//////////////////////////////////////////////////////////////
+//      Graph representation for both cases
+//
+//               3
+//              / .
+//             /   .
+//            /     .
+//           2       4
+//          /         \
+//         1           5
+//
+// Explanation:
+// 3 - major node both for simple and complex case:
+//  a) for simple case it includes via encapsulates header3
+//  b) for complex case header4 is copied into header3
+//////////////////////////////////////////////////////////////
+
+
+//////////////////////////////////////////////////////////////
+// Simple case
+bob_generate_source {
+    name: "encapsulation_source1",
+
+    srcs: ["srcs/header1.h"],
+    out: ["header1.h"],
+
+    cmd: "cp ${in} ${out}",
+    export_gen_include_dirs: ["."],
+}
+
+bob_generate_source {
+    name: "encapsulation_source2",
+    module_deps: ["encapsulation_source1"],
+
+    srcs: ["srcs/header2.h"],
+    out: ["header2.h"],
+
+    cmd: "cp ${in} ${out}",
+    export_gen_include_dirs: ["."],
+    encapsulates: ["encapsulation_source1"],
+}
+
+bob_generate_source {
+    name: "encapsulation_source3",
+    module_deps: ["encapsulation_source2"],
+
+    srcs: ["srcs/header3.h"],
+    out: ["header3.h"],
+
+    cmd: "cp ${in} ${out}",
+    export_gen_include_dirs: ["."],
+    encapsulates: ["encapsulation_source2"],
+}
+
+// Simple case
+//////////////////////////////////////////////////////////////
+
+//////////////////////////////////////////////////////////////
+// Complex case
+bob_generate_source {
+    name: "encapsulation_source5",
+
+    srcs: ["srcs/header5.h"],
+    out: ["header5.h"],
+
+    cmd: "cp ${in} ${out}",
+}
+
+bob_generate_source {
+    name: "encapsulation_source4",
+    module_deps: ["encapsulation_source5"],
+
+    srcs: ["srcs/header4.h"],
+    out: ["header4.h"],
+
+    cmd: "cp ${in} ${out}",
+    export_gen_include_dirs: ["."],
+    encapsulates: ["encapsulation_source5"],
+}
+
+bob_generate_source {
+    name: "encapsulation_source3_complex",
+    module_deps: [
+        "encapsulation_source2",
+        "encapsulation_source4",
+    ],
+
+    srcs: ["srcs/header3.h"],
+    out: ["header3.h"],
+
+    cmd: "cat ${in} ${encapsulation_source4_out} >> ${out}",
+    export_gen_include_dirs: ["."],
+    encapsulates: ["encapsulation_source2"],
+}
+
+// Complex case
+//////////////////////////////////////////////////////////////
+
+bob_binary {
+    name: "validate_source_encapsulation_simple",
+    generated_headers: ["encapsulation_source3"],
+    srcs: ["simple.c"],
+}
+
+bob_binary {
+    name: "validate_source_encapsulation_complex",
+    generated_headers: ["encapsulation_source3_complex"],
+    srcs: ["complex.c"],
```

Comment:
OK, so this test case *should* get source 1, 2, 3 gen include dirs, as well as 4's, but *not* 5's. 

Could we test this by adding a `build_wrapper` script, which looks for `-I` flags in the command line and checks that none of them contain `source5`?